EKF/ekf.cpp

```diff
@@ -91,29 +137,45 @@ bool Ekf::update()
 		predictCovariance();
 	}
 
+	// control logic
+	controlFusionModes();
+
 	// measurement updates
 
+	// Fuse magnetometer data using the selected fuson method and only if angular alignment is complete
 	if (_mag_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_mag_sample_delayed)) {
-		//fuseHeading();
-                fuseMag();
+		if (_control_status.flags.mag_3D && _control_status.flags.angle_align) {
+			fuseMag();
+
+			if (_control_status.flags.mag_dec) {
+				// TODO need to fuse synthetic declination measurements if there is no GPS or equivalent aiding
+				// otherwise heading will slowly drift
+			}
+		} else if (_control_status.flags.mag_hdg && _control_status.flags.angle_align) {
+			fuseHeading();
+		}
 	}
 
 	if (_baro_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_baro_sample_delayed)) {
 		_fuse_height = true;
 	}
 
-	if (_gps_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_gps_sample_delayed)) {
+	// If we are using GPs aiding and data has fallen behind the fusion time horizon then fuse it
+	// if we aren't doing any aiding, fake GPS measurements at the last known position to constrain drift
+	// Coincide fake measurements with baro data for efficiency with a minimum fusion rate of 5Hz
+	if (_gps_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_gps_sample_delayed) && _control_status.flags.gps) {
 		_fuse_pos = true;
 		_fuse_vert_vel = true;
 		_fuse_hor_vel = true;
 
-	} else if (_time_last_imu - _time_last_gps > 2000000 && _time_last_imu - _time_last_fake_gps > 70000) {
-		// if gps does not update then fake position and horizontal veloctiy measurements
-		_fuse_hor_vel = true;	// we only fake horizontal velocity because we still have baro
-		_gps_sample_delayed.vel.setZero();
+	} else if (!_control_status.flags.gps && !_control_status.flags.opt_flow
+		   && ((_time_last_imu - _time_last_fake_gps > 2e5) || _fuse_height)) {
+		_fuse_pos = true;
```

Comment:
@priseborough Would there be any benefit it fusing zero horizontal velocity measurements?